components/cloud/azure/src/token_credentials_with_cache/utils.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::{Arc, Mutex, RwLock};
+
+use azure_core::auth::TokenResponse;
+use time::OffsetDateTime;
+
+/// Refresh time to use in seconds
+pub(crate) const DEFAULT_REFRESH_TIME: i64 = 300;
+/// Timeout on refreshing token in seconds
+const DEFAULT_REFRESH_TOKEN_TIMEOUT: i64 = 10;
+
+fn config_default_refresh_token_timeout() -> i64 {
+    fail::fail_point!("on_config_default_refresh_token_timeout", |_| 1);
+    DEFAULT_REFRESH_TOKEN_TIMEOUT
+}
+
+#[derive(Clone)]
+pub(crate) struct TokenCache {
+    /// Cached token
+    cached_token: Arc<RwLock<Option<TokenResponse>>>,
+    /// Waited to be refreshed
+    wait_refreshed: Arc<Mutex<bool>>,
+}
+
+impl Default for TokenCache {
+    fn default() -> Self {
+        Self {
+            cached_token: Arc::new(RwLock::new(None)),
+            wait_refreshed: Arc::new(Mutex::new(false)),
+        }
+    }
+}
+
+impl TokenCache {
+    fn is_fresh(&self) -> bool {
+        fail::fail_point!("on_is_fresh_return_true", |_| { true });
+        if let Some(token) = self.cached_token.read().unwrap().as_ref() {
+            let current_time = OffsetDateTime::now_utc().unix_timestamp();
+            let expiry_time = current_time + DEFAULT_REFRESH_TIME;
+            return expiry_time <= token.expires_on.unix_timestamp();
+        }
+        false
+    }
+
+    pub fn need_update_token(&self) -> bool {
+        let now = OffsetDateTime::now_utc().unix_timestamp();
+        loop {
+            if self.is_fresh() {
+                return false;
+            }
+            if let Ok(ref mut mtx) = self.wait_refreshed.lock() {
+                // Only allowed he first one who gets the lock to refresh the token.
+                if !**mtx {
+                    **mtx = true;
```

Comment:
will here be `self.if_fresh()` is also true, according to function `update_token`?